rust/lance-table/src/io/commit/external_manifest.rs

```diff
@@ -76,6 +76,50 @@ pub struct ExternalManifestCommitHandler {
     pub external_manifest_store: Arc<dyn ExternalManifestStore>,
 }
 
+impl ExternalManifestCommitHandler {
+    /// The manifest is considered committed once the staging manifest is writen
+    /// to object store and that path is committed to the external store.
+    ///
+    /// However, to fully complete this, the staging manifest should be materialized
+    /// into the final path, the final path should be committed to the external store
+    /// and the staging manifest should be deleted. These steps may be completed
+    /// by any number of readers or writers, so care should be taken to ensure
+    /// that the manifest is not lost nor any errors occur due to duplicate
+    /// operations.
+    async fn finalize_manifest(
+        &self,
+        base_path: &Path,
+        staging_manifest_path: &Path,
+        version: u64,
+        store: &dyn OSObjectStore,
+    ) -> std::result::Result<Path, Error> {
+        // step 1: copy the manifest to the final location
+        let final_manifest_path = manifest_path(base_path, version);
+        match store
+            .copy(staging_manifest_path, &final_manifest_path)
+            .await
+        {
+            Ok(_) => {}
+            Err(ObjectStoreError::NotFound { .. }) => return Ok(final_manifest_path), // Another writer beat us to it.
+            Err(e) => return Err(e.into()),
+        };
+
+        // step 2: flip the external store to point to the final location
+        self.external_manifest_store
+            .put_if_exists(base_path.as_ref(), version, final_manifest_path.as_ref())
+            .await?;
```

Comment:
It looks like you have `store.copy` and then `external_manifest_store.put_if_exists`.

However, in the original code blocks it looks like it was `store.copy` and then `store.rename` and then `external_manifest_store.put_if_exists`.  What happened to the `rename`?

Reply:
I'm assuming that I could do this given the TODO:

```
// TODO: remove copy-rename once we upgrade object_store crate
```

This is likely from:

 * https://github.com/apache/arrow-rs/issues/4758
 * https://github.com/apache/arrow-rs/issues/5503